src/config/compiler.rs

```diff
@@ -78,7 +80,142 @@ pub(super) fn expand_macros(config: &mut Config) -> Result<(), Vec<String>> {
     if !errors.is_empty() {
         Err(errors)
     } else {
-        config.expansions = expansions;
-        Ok(())
+        Ok(expansions)
+    }
+}
+
+/// Expand trailing `*` wildcards in input lists
+fn expand_wildcards(config: &mut ConfigBuilder) {
+    let candidates = config
+        .sources
+        .keys()
+        .chain(config.transforms.keys())
+        .cloned()
+        .collect::<Vec<String>>();
+
+    for (name, transform) in config.transforms.iter_mut() {
+        expand_wildcards_inner(&mut transform.inputs, name, &candidates);
+    }
+
+    for (name, sink) in config.sinks.iter_mut() {
+        expand_wildcards_inner(&mut sink.inputs, name, &candidates);
+    }
+}
+
+fn expand_wildcards_inner(inputs: &mut Vec<String>, name: &str, candidates: &[String]) {
+    let raw_inputs = inputs.drain(..).collect::<Vec<_>>();
+    for raw_input in raw_inputs {
+        if raw_input.ends_with('*') {
+            let prefix = &raw_input[0..raw_input.len() - 1];
+            for input in candidates {
+                if input.starts_with(prefix) && input != name {
+                    inputs.push(input.clone())
+                }
+            }
+        } else {
+            inputs.push(raw_input);
+        }
+    }
+}
```

Comment:
By breaking out the actual expansion, I think this can be simplified into some straighter iteration, which suggests the input doesn't need to be `mut` at all:
```suggestion
fn expand_wildcards_inner(inputs: &mut Vec<String>, name: &str, candidates: &[String]) {
    inputs = inputs.iter()
        .map(|raw_input| expand_wildcards(&raw_input, name, candidates))
        .flatten()
        .collect();
}

// It would be nice to avoid the extra Vec allocation here, but it's not performance critical
fn expand_wildcard(raw_input: &String, name: &str, candidates: &[String]) -> Vec<String> {
    if raw_input.ends_with('*') {
        let prefix = &raw_input[0..raw_input.len() - 1];
        candidates
            .iter()
            .filter(|input| input.starts_with(prefix) && input != name)
            .cloned()
            .collect()
    } else {
        vec![raw_input.into()]
    }
}
```

Reply:
So I played around with this a bit and I'm honestly not finding much that I like better than the current state. What you have here is _almost_ better, but it starts to get awkward with borrowing and mutability when you do `inputs = inputs.iter()...`. I think what we have is good enough as-is, so I'm going to resist the temptation to spend any more time on it :sweat_smile:

Reply:
Fair enough